File: python/sort/quick.py
# This is Quick sort.
# Worst: O(n^2)
# Best: O(nlogn)
# Average: O(nlogn)

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[0]
    # arr[1:] : all elements except pivot
    low = [i for i in arr[1:] if i <= pivot]
    high = [i for i in arr[1:] if i > pivot]
    
    return quick_sort(low) + [pivot] + quick_sort(high)

# There are two algorithms of partition.

# Lomuto partition (simple)
def lomuto_partition(arr, low, high):
    pivot = arr[high]
    i = low
    for j in range(low, high):
        if arr[j] <= pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[high] = arr[high], arr[i]
    return i

# Hoare partition (few swap and stable)
def hoare_partition(arr, low, high):
    pivot = arr[(low + high) // 2]
    i = low - 1
    j = high + 1
    while True:
        i += 1
        while arr[i] < pivot:
            i += 1
        j -= 1
        while arr[j] > pivot:
            j -= 1
        if i >= j:
            return j
        arr[i], arr[j] = arr[j], arr[i]

# This quick sort is in-place with Hoare partition.
def quick_sort_inplace(arr, low, high):
    if low >= high:
        return
    
    p = hoare_partition(arr, low, high)
    quick_sort_inplace(arr, low, p)
    quick_sort_inplace(arr, p+1, high)

File: rust/sort/quick.rs
// This is Quick sort.
// Worst: O(n^2)
// Best: O(nlogn)
// Average: O(nlogn)

// There are two algorithms of partition

// Lomuto Partition (simple)
fn lomuto<T: Ord>(arr: &mut [T]) -> usize {
    let high = arr.len() - 1;
    let mut i = 0;
    for j in 0..high {
        if arr[j] <= arr[high] {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, high);
    i
}

// Hoare Partition (few swap)
fn hoare<T: Ord + Copy>(arr: &mut [T]) -> usize {
    let pivot = arr[arr.len() / 2];
    let mut i = 0;
    let mut j = arr.len() - 1;

    loop {
        while arr[i] < pivot { i += 1; }
        while arr[j] > pivot { j -= 1; }
        if i >= j { return j; }
        arr.swap(i, j);
        i += 1;
        j -= 1;
    }
}

//  This quick sort is in-place with Hoare partition.
fn quick_sort<T: Ord + Copy>(arr: &mut [T]) {
    if arr.len() <= 1 { return; }

    let p = hoare(arr); 

    let (left, right) = arr.split_at_mut(p + 1);
    quick_sort(left);
    quick_sort(right);
}